src/rabbit_amqp10_shovel.erl

```diff
@@ -353,21 +353,66 @@ add_forward_headers(#{dest := #{cached_forward_headers := Props}}, Msg) ->
 add_forward_headers(_, Msg) -> Msg.
 
 set_message_properties(Props, Msg) ->
-    maps:fold(fun(delivery_mode, 2, M) ->
-                      amqp10_msg:set_headers(#{durable => true}, M);
-                 (content_type, Ct, M) ->
-                      amqp10_msg:set_properties(
-                        #{content_type => rabbit_data_coercion:to_binary(Ct)}, M);
-                 (_, _, M) -> M
-              end, Msg, Props).
+    %% this is effectively special handling properties from amqp 0.9.1
+    maps:fold(
+      fun(content_type, Ct, M) ->
+              amqp10_msg:set_properties(
+                #{content_type => to_binary(Ct)}, M);
+         (content_encoding, Ct, M) ->
+              amqp10_msg:set_properties(
+                #{content_encoding => to_binary(Ct)}, M);
+         (delivery_mode, 2, M) ->
+              amqp10_msg:set_headers(#{durable => true}, M);
+         (correlation_id, Ct, M) ->
+              amqp10_msg:set_properties(#{correlation_id => to_binary(Ct)}, M);
+         (reply_to, Ct, M) ->
+              amqp10_msg:set_properties(#{reply_to => to_binary(Ct)}, M);
+         (message_id, Ct, M) ->
+              amqp10_msg:set_properties(#{message_id => to_binary(Ct)}, M);
+         (timestamp, Ct, M) ->
+              amqp10_msg:set_properties(#{creation_time => Ct}, M);
+         (user_id, Ct, M) ->
+              amqp10_msg:set_properties(#{user_id => Ct}, M);
+         (headers, Headers0, M) when is_list(Headers0) ->
+              %% AMPQ 0.9.1 are added as applicatin properties
+              %% TODO: filter headers to make safe
+              Headers = lists:foldl(
+                          fun ({K, _T, V}, Acc) ->
+                                  case is_amqp10_compat(V) of
+                                      true ->
+                                          Acc#{to_binary(K) => V};
+                                      false ->
+                                          Acc
+                                  end
+                          end, #{}, Headers0),
+              amqp10_msg:set_application_properties(Headers, M);
+         (Key, Value, M) ->
+              case is_amqp10_compat(Value) of
+                  true ->
+                      amqp10_msg:set_application_properties(
+                        #{to_binary(Key) => Value}, M);
+                  false ->
+                      M
+              end
+      end, Msg, Props).
 
 gen_unique_name(Pre0, Post0) ->
     Pre = rabbit_data_coercion:to_binary(Pre0),
     Post = rabbit_data_coercion:to_binary(Post0),
     Id = bin_to_hex(crypto:strong_rand_bytes(8)),
     <<Pre/binary, <<"_">>/binary, Id/binary, <<"_">>/binary, Post/binary>>.
 
+to_binary(X) ->
+    rabbit_data_coercion:to_binary(X).
```

Comment:
Why not import this function?

Reply:
I never import as it doesn't work with my code navigation :)